File: data-structure/BinaryTree/objective-c/heap/BinaryHeap.m
//
//  BinaryHeap.m
//  BinaryTree
//
//  Created by Ankui on 8/2/20.
//  Copyright © 2020 Ankui. All rights reserved.
//

#import "BinaryHeap.h"
#import "BinaryTreePrintHandler.h"


@interface BinaryHeap () <BinaryTreeProtocol> {
    NSUInteger _capacity;
    NSMutableArray *_elements;
}
@end

@implementation BinaryHeap

static const NSUInteger DEFAULT_CAPACITY = 10;

- (instancetype)init {
    if (self = [super init]) {
        _capacity = DEFAULT_CAPACITY;
        _elements = [NSMutableArray arrayWithCapacity:DEFAULT_CAPACITY];
    }
    return self;
}

+ (instancetype)heapWithArray:(NSArray *)array {
    BinaryHeap *heap = [self heapWithArray:array comparator:nil];
    return heap;
}
+ (instancetype)heapWithArray:(NSArray *)array comparator:(id<HeapComparator> _Nullable)comparator {
    BinaryHeap *heap = [BinaryHeap heapWithComparator:comparator];
    if (array.count) {
        heap->_size = array.count;
        NSUInteger capacity = MAX(heap->_capacity, array.count);
        heap->_elements = [NSMutableArray arrayWithCapacity:capacity];
        for (NSUInteger i = 0; i < array.count; i++) {
            [heap->_elements addObject:array[i]];
        }
        [heap _heapify];
    }
    return heap;
}
+ (instancetype)heapWithArray:(NSArray *)array block:(NSInteger (^)(id _Nullable, id _Nullable))block {
    BinaryHeap *heap = [BinaryHeap heapWithBlock:block];
    if (array.count) {
        heap->_size = array.count;
        heap->_capacity = MAX(heap->_capacity, array.count);
        heap->_elements = [NSMutableArray arrayWithCapacity:heap->_capacity];
        for (NSUInteger i = 0; i < array.count; i++) {
            [heap->_elements addObject:array[i]];
        }
        [heap _heapify];
    }
    return heap;
}
- (void)clear {
    if (_size == 0) { return; }
    [_elements removeAllObjects];
    _size = 0;
}
- (void)add:(id)element {
    [self _elementCheck:element];
    [self _ensureCapacity:_size + 1];
    _elements[_size++] = element;
    [self _siftUp:_size - 1];
}
- (id)get {
    [self _emptyCheck];
    return _elements.firstObject;
}
- (id)remove {
    [self _emptyCheck];
    id top = _elements.firstObject;
    _elements[0] = _elements[_size - 1];
    _size--;
    [_elements removeLastObject];
    [self _siftDown:0];
    return top;
}
- (id)replace:(id)element {
    [self _elementCheck:element];
    id top = _elements.firstObject;
    if (_size == 0) {
        _elements[0] = element;
        _size++;
    } else {
        _elements[0] = element;
        [self _siftDown:0];
    }
    return top;
}
- (void)print {
    [BinaryTreePrintHandler println:self];
    printf("\n-----------------------------\n");
}

- (void)_heapify {
//    for (NSUInteger i = 0; i < _size; i++) {
//        [self _siftUp:i];
//    }
    
    for (NSInteger i = (_size >> 1) - 1; i >= 0; i--) {
        [self _siftDown:i];
    }
}

- (void)_siftDown:(NSUInteger)index {
    if (_elements.count == 0) { return; }
    id element = _elements[index];
    NSUInteger half = _size >> 1;
    while (index < half) {
        NSUInteger lIndex = (index << 1) + 1;
        id left = _elements[lIndex];
        NSUInteger rIndex = lIndex + 1;
        if (rIndex < _size && [self compare:_elements[rIndex] e2:left] > 0) {
            left = _elements[lIndex = rIndex];
        }
        if ([self compare:element e2:left] >= 0) { break; }
        _elements[index] = left;
        index = lIndex; // CARE!!!
    }
    _elements[index] = element;
}
- (void)_siftUp:(NSUInteger)index {
    id element = _elements[index];
    while (index > 0) {
        NSUInteger pIndex = (index - 1) >> 1;
        id parent = _elements[pIndex];
        if ([self compare:element e2:parent] <= 0) { break; }
        _elements[index] = parent;
        index = pIndex; // CARE!!!
    }
    _elements[index] = element;
}
- (void)_ensureCapacity:(NSInteger)capacity {
    if (_capacity >= capacity) { return; }
    _capacity = _capacity + (_capacity >> 1);
    NSMutableArray *oldElements = [NSMutableArray arrayWithArray:_elements];
    _elements = [NSMutableArray arrayWithCapacity:_capacity];
    for (NSUInteger i = 0; i < _size; i++) {
        _elements[i] = oldElements[i];
    }
}
- (void)_emptyCheck {
    if (_size == 0) {
        assert("Heap is empty");
    }
}
- (void)_elementCheck:(id)element {
    if (!element) {
        assert("element msut not be nil");
    }
}


// MARK: - BinaryTreeProtocol
- (id)root {
    return nil;
}
- (id)left:(id)object {
    NSInteger index = [object integerValue];
    index = (index << 1) + 1;
    return index >= _size ? nil : @(index);
}
- (id)right:(id)object {
    NSInteger index = [object integerValue];
    index = (index << 1) + 2;
    return index >= _size ? nil : @(index);
}
- (id)string:(id)object {
    NSInteger index = [object integerValue];
    return _elements[index];
}
@end

File: data-structure/data-structure/heap/binary_heap.c
//
//  binary_heap.c
//  data-structure
//
//  Created by Ankui on 8/2/20.
//  Copyright © 2020 Ankui. All rights reserved.
//

#include "binary_heap.h"

int heap_size_ = 0;
int heap_capacity = 0;
int *heap_elements = NULL;

// MARK: - private
void heap_empty_check(void) {
    if (heap_size_ == 0) {
        assert("Heap is empty");
    }
}
void heap_ensure_capacity(int capacity) {
    if (heap_capacity >= capacity) { return; }
    
    heap_capacity = heap_capacity + (heap_capacity >> 1);
    int *newElements = malloc(sizeof(int) * heap_capacity);
    for (int i = 0; i < heap_size_; i++) {
        newElements[i] = heap_elements[i];
    }
    heap_elements = newElements;
    free(newElements);
}
void heap_sift_up(int index) {
    int e = heap_elements[index];
    while (index > 0) {
        int pIndex = (index - 1) >> 1;
        int p = heap_elements[pIndex];
        if (e <= p) { break;; }
        heap_elements[index] = p;
        index = pIndex;
    }
    heap_elements[index] = e;
    
//    while (index > 0) {
//        int pIndex = (index - 1) >> 1;
//        int p = heap_elements[pIndex];
//        if (e <= p) { return; }
//        int tmp = heap_elements[index];
//        heap_elements[index] = heap_elements[pIndex];
//        heap_elements[pIndex] = tmp;
//        index = pIndex;
//    }
}
void heap_sift_down(int index) {
    int e = heap_elements[index];
    int half = heap_size_ >> 1;
    while (index < half) {
        int lIndex = (index << 1) + 1;
        int left = heap_elements[lIndex];
        int rIndex = lIndex + 1;
        if (rIndex < heap_size_ && heap_elements[rIndex] > left) {
            left = heap_elements[lIndex = rIndex];
        }
        if (e >= left) { break; }
        heap_elements[index] = left;
        index = lIndex;
    }
    heap_elements[index] = e;
}

void heap_heapify(void) {
//    for (int i = 1; i < heap_size_; i++) {
//        heap_sift_up(i);
//    }
    
    for (int i = (heap_size_ >> 1) - 1; i >= 0; i--) {
        heap_sift_down(i);
    }
}


void create_heap_list(int* arr, int size) {
    if (size == 0) {
        heap_capacity = DEFAULT_CAPACITY;
        heap_elements = malloc(sizeof(int) * heap_capacity);
    } else {
        heap_size_ = size;
        heap_capacity = fmax(DEFAULT_CAPACITY, size);
        heap_elements = malloc(sizeof(int) * heap_capacity);
        for (int i = 0; i < size; i++) {
            heap_elements[i] = arr[i];
        }
        heap_heapify();
    }
}

void create_heap(void) {
    heap_capacity = DEFAULT_CAPACITY;
    heap_elements = malloc(sizeof(int) * heap_capacity);
}
int heap_size(void) {
    return heap_size_;
}
bool heap_isEmpty(void) {
    return heap_size_ == 0;
}
void heap_clear(void) {
    if (heap_size_ == 0) { return; }
    free(heap_elements);
    heap_size_ = 0;
}
void heap_add(int element) {
    heap_ensure_capacity(heap_size_ + 1);
    heap_elements[heap_size_++] = element;
    heap_sift_up(heap_size_ - 1);
}
int heap_get(void) {
    heap_empty_check();
    return heap_elements[0];
}
int heap_remove(void) {
    heap_empty_check();
    int top = heap_elements[0];
    int lastIndex = --heap_size_;
    heap_elements[0] = heap_elements[lastIndex];
    heap_elements[lastIndex] = -1;
    heap_sift_down(0);
    return top;
}
int heap_replace(int element) {
    int top = -1;
    if (heap_size_ == 0) {
        heap_elements[0] = element;
        heap_size_++;
    } else {
        top = heap_elements[0];
        heap_elements[0] = element;
        heap_sift_down(0);
    }
    return top;
}
